### PR TITLE
RainForests apply calibration skeleton

### PR DESCRIPTION
**Pull request by benowen-bom**
This PR adds a skeleton for RainForest `apply_rainforests_calibration` CLI and associated plugin. 

Future PRs following on from this will add in functionality and tests, however following the review of the CLI interface through this PR the interface is not expected to change in subsequent PRs.

Open questions (@tjtg, I'd be interested in your thoughts):
- ~~Additional packages are required RainForest calibration. Which envs should these be added to (currently they have been added to all envs except `conda-forge.yml`)?~~
- ~~Where's the best place to call `initialise_model_config`? Currently being called in the CLI prior to calling process. Other options could be that it is called in the `__init__` for `ApplyRainForestsCalibration` or within the associated process function, however both these options will change the interface for the process function.~~

Still to-do:
- [x] Add test for ~~`initialise_model_config`~~ `__init__`.

Testing:
 - [x] Ran tests and they passed OK
 - [x] Added new tests for the new feature(s)

### CODE REVIEW

**Comment by tjtg** [comment]
> Additional packages are required RainForest calibration. Which envs should these be added to (currently they have been added to all envs except `conda-forge.yml`)?

I'd suggest adding these packages to environment_b and latest.

The current [CI error on the conda-forge environment](https://github.com/metoppv/improver/runs/6251837931?check_suite_focus=true#step:6:421) will need the lightgbm import to be moved inside the functions where it is used rather than being at the top of the file. This is similar to other optional packages in IMPROVER.

Please also update the dependencies documentation in [doc/source/Dependencies.rst](https://github.com/metoppv/improver/blob/master/doc/source/Dependencies.rst)

> Where's the best place to call `initialise_model_config`? Currently being called in the CLI prior to calling process. Other options could be that it is called in the `__init__` for `ApplyRainForestsCalibration` or within the associated process function, however both these options will change the interface for the process function.

The contents of `initialise_model_config` look like they should become the `__init__` method of `ApplyRainForestsCalibration`. Many other IMPROVER plugin classes don't have anything useful to do in the `__init__` method, so `__init__` is either tiny or doesn't exist. This plugin does have some useful setup that can be done in the `__init__` method. I don't think there's any other realistic use of `initialise_model_config` outside this class, so no reason to have it at top level outside the classes.

File: improver/calibration/rainforest_calibration.py
Comment (tjtg) [inline improver/calibration/rainforest_calibration.py:154]: I assume feature cubes are expected to have the same dimensions as the forecast cube, but represent different parameters.

File: improver_tests/calibration/rainforests_calibration/test_ApplyRainForestsCalibration.py
Comment (tjtg) [inline improver_tests/calibration/rainforests_calibration/test_ApplyRainForestsCalibration.py:78]: There's currently only one of these `not TREELITE_ENABLED` tests. If this grows in the future, consider moving these treelite tests to a separate file and using `pytest.importorskip()` similar to the existing lightgbm above.

File: improver/calibration/rainforest_calibration.py
Comment (tjtg) [inline improver/calibration/rainforest_calibration.py:103]: Suggest using an "ask for forgiveness rather than permission" approach here.
```suggestion
        try:
            from treelite_runtime import Predictor
            self.treelite_enabled = True
        except ModuleNotFoundError:
            warnings.warn(
                "Module treelite_runtime unavailable. Defaulting to using lightgbm Boosters."
            )
            self.treelite_enabled = False
```

Comment (tjtg) [reply improver/calibration/rainforest_calibration.py:103]: There's another similar usage in the unit test code.

**Review by tjtg — COMMENTED** [review]
This is a good start on adding the RainForests calibration into IMPROVER. The comments I've made are a bunch of minor items and should be pretty easy to address.
It would be worthwhile getting a second review of this PR as it's adding a new CLI.

File: improver/calibration/rainforest_calibration.py
Comment (benowen-bom) [reply improver/calibration/rainforest_calibration.py:45]: No worries. I've dropped this class definition.

File: improver/calibration/rainforest_calibration.py
Comment (benowen-bom) [reply improver/calibration/rainforest_calibration.py:83]: Yeah, this was a carry over of when the previous function was converted into the class __init__. I'll update the docstring as suggested.

File: improver/calibration/rainforest_calibration.py
Comment (benowen-bom) [reply improver/calibration/rainforest_calibration.py:59]: I'll double check this, but as it stands the docstring is inconsistent as you've highlighted. I'll update the docstring, or fix the intended use to align with the above.

File: improver/calibration/rainforest_calibration.py
Comment (benowen-bom) [reply improver/calibration/rainforest_calibration.py:153]: Yeah, that is correct. I'll update the docstrings to explicitly state this.

File: improver/calibration/rainforest_calibration.py
Comment (benowen-bom) [reply improver/calibration/rainforest_calibration.py:170]: I've updated this to return the forecast cube as-is, with an comment to explicitly state that this is the intended behaviour and why.

File: improver/calibration/rainforest_calibration.py
Comment (benowen-bom) [reply improver/calibration/rainforest_calibration.py:154]: Yeah, though there is the possibility here that some features (derived fields) can have no realization dimension. In processing, these cubes are broadcast to match the cubes that do contain realization dimension.

I'll update the docstring to describe this behaviour in more detail.

**Review by MoseleyS — CHANGES_REQUESTED** [review]

File: improver/calibration/rainforest_calibration.py
Comment (benowen-bom) [reply improver/calibration/rainforest_calibration.py:154]: I've updated the docstring to be more explicit; hopefully this is clearer now, but would appreciate any suggestions.

File: improver/cli/apply_rainforests_calibration.py
Comment (benowen-bom) [reply improver/cli/apply_rainforests_calibration.py:72]: Good to know! I've replaced add instances of `nthreads` with `threads`.

File: improver_tests/calibration/rainforests_calibration/test_ApplyRainForestsCalibration.py
Comment (benowen-bom) [reply improver_tests/calibration/rainforests_calibration/test_ApplyRainForestsCalibration.py:78]: That sounds like a good idea. For now I might leave this as is, but will keep this in mind in follow-up PR(s) which add the functionality to the Plugin.

File: improver/calibration/rainforest_calibration.py
Comment (benowen-bom) [reply improver/calibration/rainforest_calibration.py:103]: Good suggestion! I've updated both instances as suggested.

File: improver/calibration/rainforest_calibration.py
Comment (benowen-bom) [reply improver/calibration/rainforest_calibration.py:59]: The intention of this parameter is the number of threads to use for prediction processing, rather than initialisation. So these parameters are being used as intended, but the docstring didn't reflect this. I've updated the docstring for `threads` to:
```
threads:
    Number of threads to use during prediction with tree-model objects.
```

File: improver/calibration/rainforest_calibration.py
Comment (benowen-bom) [reply improver/calibration/rainforest_calibration.py:115]: The default behaviour for `dict.get` is to return `None` when no default value is provided (https://docs.python.org/3/library/stdtypes.html#dict.get). I could update the `.get` calls to be more explicit, but I think as it stands the code should not produce a `KeyError`.

The intention here is that in order to use treelite `Predictor` for the tree_model, all error thresholds must have a path specified for `treelite_model` otherwise it will default to using lightgbm `Booster` instead. I believe the code does this currently.

However, looking over this I think the situation that has not been addressed if where a `treelite_model` is specified, but not the corresponding `lightgbm_model`. In this case the code will fail as a `None` value will be passed into the `Booster` initialisation. I can add a check to ensure there are no `None` values in `lightgbm_model_filenames` before loading to catch this situation.

File: improver/calibration/rainforest_calibration.py
Comment (tjtg) [reply improver/calibration/rainforest_calibration.py:148]: I think this is pretty clear now.

File: improver/cli/apply_rainforests_calibration.py
Comment (benowen-bom) [reply improver/cli/apply_rainforests_calibration.py]: Good suggestion. I've added a type conversion when passing features through to the process function.

File: improver/calibration/rainforest_calibration.py
Comment (benowen-bom) [reply improver/calibration/rainforest_calibration.py:62]: Fair question. I should add that this is my first time using Mock/MonkeyPatch and it took some effort to get my head around what/how to test the __int__ function, so I completely understand the opacity.

The intention with the unit tests is to test that the relevant model objects are being initialised given the values passed via model_config; for the test we are not really interested in the result model tree initialisation, just that the relevant load function is being used. Consequently, the goal of mocking here is to override the lightgbm Booster and treelite Predictor __init__ functions to generate Mock instances of these objects. In doing this, we can test that the relevant load function is being called without actually having to load anything, which simplifies the tests as there is no need to create sample files to load.

So the tests should pass, despite the files that would be loaded not existing.

File: improver/calibration/rainforest_calibration.py
Comment (benowen-bom) [reply improver/calibration/rainforest_calibration.py:98]: I've moved where class variable assignment `self.error_thresholds` (and where the conversion is done) to earlier in the function, and added a comment to try and make this clearer.

File: improver/calibration/rainforest_calibration.py
Comment (MoseleyS) [inline improver/calibration/rainforest_calibration.py:108]: You don't need `error_thresholds` (and I don't like using the same name twice for different contents) if you use the `values()` method of the dictionary.
```suggestion
        self.error_thresholds = np.array(
            list(model_config_dict.keys()), dtype=np.float32
        )

        lightgbm_model_filenames, treelite_model_filenames = [
            [v.get(k) for v in model_config_dict.values()]
            for k in ["lightgbm_model", "treelite_model"]
        ]
```

File: improver/calibration/rainforest_calibration.py
Comment (MoseleyS) [reply improver/calibration/rainforest_calibration.py:109]: My mistake. I was wrong on the default behaviour of `get`. I'm now wondering if the behaviour has changed, or if I was always wrong! Never mind...

You are right in that the behaviour does not trap incomplete `lightgbm_model` file names, so the process should fail somewhere else. I don't know where it will fail as your tests don't cover this. I suggest you add a test for this, then modify the behaviour if necessary to return a useful Exception.

File: improver/calibration/rainforest_calibration.py
Comment (MoseleyS) [reply improver/calibration/rainforest_calibration.py:62]: Thanks for the explanation - I thought that might be what was happening, but you've made it a lot clearer for me!
What this method prevents, is testing that the true objects do initiate, and what happens when one of the file names is `None` (see separate comment). I think the best solution is to trap that potential `None` case so that you don't have to test the true behaviour of the external module.

**Review by MoseleyS — CHANGES_REQUESTED** [review]

File: improver/calibration/rainforest_calibration.py
Comment (benowen-bom) [reply improver/calibration/rainforest_calibration.py:62]: No worries. I tried the case where `lightgbm_model` is missing for one of the error-threshold keys and it fails at the point of initialising the associated `Booster` object as `None` is not a valid value. However, I agree that the best way to handle this is test for `None` before loading in values, so I've put in a check before initialising the `Booster` objects and raising a `ValueError` if any `None` values are in `lightgbm_model_filenames`.

I've also added a unit-test to check this case.

File: improver/calibration/rainforest_calibration.py
Comment (benowen-bom) [reply improver/calibration/rainforest_calibration.py:109]: No worries, it's always good to check these things.

As described in the next comment, I've added a check for this case and added a unit-test to check correct behaviour.

File: improver/calibration/rainforest_calibration.py
Comment (benowen-bom) [reply improver/calibration/rainforest_calibration.py:108]: Good suggestion, I'll make this change. I hadn't considered this approach and it removes an unnecessary variable.

Comment (benowen-bom) [reply improver/calibration/rainforest_calibration.py:108]: One thing that occurred to me in mulling this over is the possibility that the user passes in a set of error thresholds that are unordered. In future steps a threshold `DimCoord` is created from `error_thresholds` which will inevitably fail on `error_thresholds` being unordered.

I'll add a function to enforce ascending order to `error_thresholds` (and the associated `tree_models` accordingly).

**Review by MoseleyS — CHANGES_REQUESTED** [review]
Thanks for the updates @benowen-bom. As you'll see, I think your sorting could be simplified a lot.

One question I haven't worked out yet - why do you need to support TWO methods? 

Can I ask you not to rebase your branch each time to push new changes? It changes the history and forces me to re-review the whole PR, rather than incrementally reviewing your updates. You should only need to rebase if Github tells you there are conflicts with master, and even then, a merge is sometimes the better choice.

**Comment by tjtg** [comment]
@MoseleyS
> why do you need to support TWO methods?

[LightGBM](https://lightgbm.readthedocs.io/en/latest/) is a moderate sized library (~2MB on conda-forge) and has fairly basic dependencies - mainly scipy, numpy and scikit-learn. LightGBM contains functionality for both training GBDT models and prediction using GBDTs. 

[Treelite](https://treelite.readthedocs.io/en/latest/) is also a moderate sized library (~1.5 MB on conda-forge). Our usage has shown a significant performance enhancement with Treelite. However, Treelite only handles prediction (eg. applying the GBDTs already trained using LightGBM) and it depends on having a C compiler and toolchain available. The C compiler will already exist on many systems, but it won't on some systems and it may not be easy to install. There are also a variety of other vendor-specific GBDT acceleration solutions such as [Intel DAAL](https://intelpython.github.io/daal4py/algorithms.html#daal4py.gbt_classification_prediction) and [Nvidia cuML](https://docs.rapids.ai/api/cuml/stable/api.html#forest-inferencing) -  one of these may be preferable to treelite depending on the available hardware. It seems best to avoid a strict Treelite requirement for RainForests.

File: improver/calibration/rainforest_calibration.py
Comment (benowen-bom) [reply improver/calibration/rainforest_calibration.py:134]: Thanks @MoseleyS. I agree that it seems overly complicated.

My reasoning for going with the above was the the keys in model_config_dict will generally be strings as they will be sourced from json files. So the issue is that `sorted(model_config_dict.items())` may not produce the desired ordering (for instance, if the user provides positive values with leading `+` the sorted strings will begin with positive values first).

Ultimately, I think the best solution will be something like the above, with a preceding step to cast the keys as float32:
```
model_config_dict = {np.float32(threshold) : model_config_dict[threshold] for threshold in model_config_dict.keys()}
sorted_model_config_dict = collections.OrderedDict(sorted(model_config_dict.items()))
```

Then we simplify the assignment of `self.error_thresholds` as the keys are now float32, and we ensure that the thresholds and the associated tree-models are sorted (as you highlight above) without the need for the complicated mess in `_ordered_error_thresholds`.

File: improver/calibration/rainforest_calibration.py
Comment (benowen-bom) [reply improver/calibration/rainforest_calibration.py]: With the update to sort the thresholds, the conversion to float now occurs as the first step. The dictionary keys are converted to floats, and then the dictionary is sorted based on thresholds (key values). Hopefully this is clearer; I've added a comment to also explain the need to cast to float initially.

**Comment by benowen-bom** [comment]
> As you'll see, I think your sorting could be simplified a lot.

Thanks for the suggestion @MoseleyS. Hopefully the revised version is simpler (if anything it removes a few lines of code).

> why do you need to support TWO methods?

@tjtg has explained this better than I could, but essentially we've found that treelite provides a significant speed up in the prediction step. To build the treelite Predictors in the first place (not done here) one needs the lightgbm boosters; model training functionality (which will be added in a future PR(s)) will be done via the lightgbm library, so this is essentially the core dependency for the RainForests calibration. For reasons Tom mentioned above, we've decided not to make treelite a required dependency for using RainForests calibration.

**Comment by benowen-bom** [comment]
> Can I ask you not to rebase your branch each time to push new changes? It changes the history and forces me to re-review the whole PR, rather than incrementally reviewing your updates. You should only need to rebase if Github tells you there are conflicts with master, and even then, a merge is sometimes the better choice.

And sorry about this. I wasn't aware of this behaviour on your end (and can appreciate how annoying this would be). I'm relatively a novice when it comes to git, so feedback like this is really useful!

File: improver/calibration/rainforest_calibration.py
Comment (MoseleyS) [reply improver/calibration/rainforest_calibration.py]: I don't think this quite does everything you want. The issues I see are:

1. Modifies method argument `model_config_dict`
2. Uses `sorted_model_config_dict` to derive the list of error thresholds, but the unsorted `model_config_dict` to get the file names.
3. Dict comprehension using `keys` when `items` would be much better

I'll put in a new review to use the sorted version everywhere.

File: improver/calibration/rainforest_calibration.py
Comment (MoseleyS) [inline improver/calibration/rainforest_calibration.py:114]:
```suggestion
        sorted_model_config_dict = OrderedDict(
            sorted({np.float32(k): v for k, v in model_config_dict.items()}.items())
        )

        self.error_thresholds = np.array([*sorted_model_config_dict.keys()])

        lightgbm_model_filenames = [
            threshold_dict.get("lightgbm_model")
            for threshold_dict in sorted_model_config_dict.values()
        ]
        treelite_model_filenames = [
            threshold_dict.get("treelite_model")
            for threshold_dict in sorted_model_config_dict.values()
        ]
```

**Review by MoseleyS — CHANGES_REQUESTED** [review]
As discussed - I think the sorted dict needs to be used in place of the unsorted dict. You probably need a unit test to show that you're getting the sorting you desire too.

File: improver/calibration/rainforest_calibration.py
Comment (benowen-bom) [reply improver/calibration/rainforest_calibration.py]: Sorry about this. I was too focused on the error_thresholds and didn't think this through carefully enough. I've gone with you're suggestion below which addresses all these points.

File: improver/calibration/rainforest_calibration.py
Comment (benowen-bom) [reply improver/calibration/rainforest_calibration.py]: Thanks for this suggestion @MoseleyS. I've gone with the above and expanded the unit test `test__init__for_ordered_error_threshold` to check that the sorting is as expected and that the consistency between error threshold and tree_model is preserved.

**Comment by benowen-bom** [comment]
> As discussed - I think the sorted dict needs to be used in place of the unsorted dict.

Sorry about this @MoseleyS, I wasn't careful enough on the previous update (I clearly rushed it and was to focused on the error thresholds). I've gone with your suggested change, and updated the unit tests to ensure we get what is expected.

**Review by MoseleyS — CHANGES_REQUESTED** [review]
I'm happy with the functionality now, but now that I understand the code, I can't resist asking for the tests to be parametrized as there is a lot of repetition and overlap. To show you what I mean, I've created a PR to this PR: https://github.com/benowen-bom/improver/pull/1

**Comment by MoseleyS** [comment]
I hope you don't mind me making so many suggestions! I also hope you agree that the end result is better code! Our working hours don't seem to overlap, so it is taking longer to progress than is normal. I think the end is in sight though.

**Comment by benowen-bom** [comment]
> I'm happy with the functionality now, but now that I understand the code, I can't resist asking for the tests to be parametrized as there is a lot of repetition and overlap. To show you what I mean, I've created a PR to this PR: [benowen-bom#1](https://github.com/benowen-bom/improver/pull/1)

Thanks @MoseleyS for this suggestion. I think this does a fantastic job at removing the repetition in the tests and means that the same unit test is applied for calls to `__init__`, so I have adopted your change, with a couple of minor tweaks. These changes are to the testing regime, rather than the test itself and relate to running tests when the treelite dependency is missing.

**Comment by benowen-bom** [comment]
> I hope you don't mind me making so many suggestions! I also hope you agree that the end result is better code!

@MoseleyS: I really appreciate your input and interest in this work! I think the quality of the code is much better through this process, and it's also been a learning experience for me. So there is no need to apologise with making so many suggestions. 

> Our working hours don't seem to overlap, so it is taking longer to progress than is normal. I think the end is in sight though.

The timeframe to get this done has been longer than I expected, but as you allude to, if time zones overlapped nicely this would have been a couple of days rather than a week.

**Review by MoseleyS — APPROVED** [review]
I'm happy with this now!

**Review by tjtg — APPROVED** [review]
I've had another check through after the review changes. I'm happy for this to go ahead and get merged.

**Comment by benowen-bom** [comment]
Thanks @tjtg and @MoseleyS!

**Pull request merged**
